validator/traverse.py

```diff
@@ -0,0 +1,469 @@
+from collections.abc import Iterator
+from dataclasses import dataclass
+from typing import Any, NamedTuple, Self
+
+from validator.bundle import RESOURCE_REF, Bundle, GraphqlField, GraphqlType
+from validator.jsonpath import JSONPath, JSONPathField, JSONPathIndex
+
+
+@dataclass(frozen=True)
+class Node:
+    bundle: Bundle
+    data: Any
+    file_schema_path: str | None
+    graphql_field_name: str | None
+    graphql_type_name: str | None
+    jsonpaths: list[JSONPath]
+    path: str
+    schema: Any
+    schema_one_of_root: Any
+    schema_path: str | None
+    parent: Self | None
+
+    @property
+    def graphql_type(self) -> GraphqlType | None:
+        if self.graphql_type_name:
+            return self.bundle.graphql_lookup.get_by_type_name(self.graphql_type_name)
+        return None
+
+    @property
+    def graphql_field(self) -> GraphqlField | None:
+        if self.graphql_field_name is None:
+            return None
+        if graphql_type := self.graphql_type:
+            return graphql_type.get_field(self.graphql_field_name)
+        return None
+
+    def is_resource_ref(self) -> bool:
+        return (
+            self.schema is not None
+            and isinstance(self.schema, dict)
+            and self.schema.get("$ref") == RESOURCE_REF
+        )
+
+    def resolve_ref(self) -> dict[str, Any] | None:
+        if self.data and isinstance(self.data, str):
+            return self.bundle.data.get(self.data)
+        return None
+
+
+def traverse_data(bundle: Bundle) -> Iterator[Node]:
+    """
+    Traverse the data files in the bundle and yield nodes representing each data item.
+
+    Each node contains information about the data, its schema, GraphQL type, and JSON paths.
+
+    Args:
+        bundle (Bundle): The bundle containing data files and schemas.
+    Yields:
+        Node: A node representing a data item, including its data, schema, GraphQL type
+    """
+    for datafile_path, datafile in bundle.data.items():
+        datafile_schema = datafile.get("$schema")
+        schema = bundle.schemas.get(datafile_schema, {}) if datafile_schema else None
+        graphql_type = (
+            bundle.graphql_lookup.get_by_schema(datafile_schema)
+            if datafile_schema
+            else None
+        )
+        schema_info = _resolve_schema(
+            schema_path=datafile_schema,
+            schema=schema,
+            bundle=bundle,
+            data=datafile,
+            graphql_type=graphql_type,
+        )
+        graphql_type_name = _get_init_graphql_type_name(bundle, graphql_type, datafile)
```

Comment:
NIT: Something like this is more readable?
```suggestion
        if data_schema:
            schema = bundle.schemas.get(datafile_schema, {})
            graphql_type = bundle.graphql_lookup.get_by_schema(datafile_schema)
            graphql_type_name = _get_init_graphql_type_name(bundle, graphql_type, datafile)
         else:
             schema = None
             graphql_type_name = None
         
         schema_info = _resolve_schema(
             schema_path=datafile_schema,
             schema=schema,
             bundle=bundle,
             data=datafile,
             graphql_type=graphql_type,
         )
```